### PR TITLE
FIX: Range option lower limit caused issues with indexing arrays

### LINKED ISSUE #221: BUG: Error plotting range gates and data if not from 0th gate (indexing issue)
# BUG 

I'm assuming this is a fairly minor issue if no user has found it or brought it up. I was going to try to fix it quickly on PR #220 but it turns out its quite pervasive as a lot of the code for plotting boundaries and the grid ect. uses the ranges values for indexing. So it's probably best it has it's own fix and testing done.
If this turns out to be a large fix for too little reward, we can state that ranges is just setting an upper limit. 

## Priority

- [x] Minor - typo, incorrect naming, fails on specific uncommon situation
- [ ] Major - misrepresentation of data, failure in a common situation 
- [ ] Urgent - failure in usage, large misrepresentation

## Information
**python version**: 3.9
**OS**: MacOS Catalina
**matplotlib version**: 4.5.1

## Example of the bug 

```python
import matplotlib.pyplot as plt
import pydarn

file = '/Users/carley/Documents/data/20211008.2000.00.rkn.a.fitacf'
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()
pydarn.Fan.plot_fan(fitacf_data, grid=True, fov_color='blue',scan_index=2, radar_label=True, groundscatter=True, ranges=[20,60], lowlat=50)
plt.show()
```

Screen Shot 2022-02-02 at 12 45 13 PM

### Attempts

The plotting problem was due to ranges being used to initialize an array of zeros, hence the plotting of the lower ranges gates going off to the 0,0 point of the plot. This is easily fixed, but boundaries, grids and fov_color all use the ranges values as indexing, this is also easily fixed by jsut adding in ranges[1]-ranges[0] for ranges[1] in most places. 
However, there is an issue where now where the fov is correct, but it's plotting data from the bottom fov which isn't zero anymore, I've not yet fixed this but can keep looking. 

### Data Location

Not data specific. 

## Potential Bug Location 

radar_fov.py and fan.py
Might have knock on effects in grid.py and future maps.py, unsure.

### PR DESCRIPTION
**Pull request by carleyjmartin**
# Scope 

This pr includes changes made in #220 as they help fix this too. So do that one first. 

This PR is to fix the issue which occurred when `ranges=[non-zero lower limit, upper limit]`
Essentially, `ranges` is used in *radar_fov, plot_fan* and *plot_fov* as if it were an index. If it doesn't start at zero we then get a bunch of zeros at the start of the array as it was initialised as a zero array in *radar_fov*, which was indexed from `[0, ranges[1]]`. This is fixed and the plot doesn't look too crazy any more. 
In *radar_fov, plot_fan* and *plot_fov*, indexing has been changed to account for a different start:end value. For example, what was written as `[0:ranges[1]]` is now `[0:ranges[1]-ranges[0]]` (converting the `ranges` values to indexes if `ranges[0]` is non-zero). Some other arrays also needed a `-range[0]` to correct their index. 
The way that we find `good_data` in the read in file has changed to use an *np.where* so we can use two conditions, where `slist` is greater than the lowest range and less than the highest range. 
You may also see that some for loops now loop from `range(ranges[1]-ranges[0])` (to get a correct index value) but some may loop over `range(ranges[0],ranges[1])` as this is looping over the actual gate numbers not the indices. 

Again, grid.py has no changes due to the fact it only uses the default values of the FOV, if this changes some loops may need updating there too. 

**issue:** to close #221 

## Approval

**Number of approvals:** 2 at least, theres quite a lot of 'thinky' changes so someone who knows the data well and can see if there's maybe some offsets or anything (like if we should be doing a -1 or +1 when looping or something, I dunno).

## Test

**matplotlib version**: 4.5.1
**Note testers: please indicate what version of matplotlib you are using**

Testing FAN:
```python
import matplotlib.pyplot as plt
import pydarn
file = '/Users/carley/Documents/data/20211008.2000.00.rkn.a.fitacf'
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()
pydarn.Fan.plot_fan(fitacf_data, grid=True, fov_color='blue',scan_index=2,
                    radar_label=True, groundscatter=True, ranges=[30,40], lowlat=50)
    
plt.show()
```
Screen Shot 2022-02-02 at 3 48 41 PM

Compared to full FOV:
Screen Shot 2022-02-02 at 3 50 03 PM

Testing some FOV:
```python
import pydarn
from datetime import datetime
import matplotlib.pyplot as plt 


fanplot=pydarn.Fan.plot_fov(66, datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True, lowlat= 50,boundary=True, line_color='red', grid = True, rsep=15, frang=500, ranges=[0,100])
fanplot1=pydarn.Fan.plot_fov(5, datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True,  ax=1, boundary=True, line_color='blue', grid = True, rsep=45, frang=100, ranges=[0,40])
fanplot2=pydarn.Fan.plot_fov(64, datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True,  ax=1, boundary=True, line_color='green', grid = True, rsep=15, frang=90)
fanplot3=pydarn.Fan.plot_fov(65, datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True,  ax=1, boundary=True, line_color='orange', grid = True)
fanplot4=pydarn.Fan.plot_fov(6, datetime(2021, 2, 5, 12, 5), radar_location=True, radar_label=True,  ax=1, boundary=True, grid = True, line_alpha=1, rsep=45, frang=0,ranges=[50,75])
plt.show()
```
Screen Shot 2022-02-02 at 3 51 36 PM

*Reminder, please check your code is:
  - copyrighted (if applicable)
  - Appropriate disclaimer added
  - Modification line filled in (if applicable)*

### CODE REVIEW

**Comment by Shirling-VT** [comment]
MacOS 10.15.7
matplotlib 3.4.2
python 3.8.0

Both the Fan and FOV test codes provide the same results as shown. Would need more time to think about if these changes make sense or not.

**Comment by mts299** [comment]
Looks good just would merge #220 first since this contains that PR. 

@billetd do you remember why you did -1 on the lower range limit?

**Review by mts299 — CHANGES_REQUESTED** [review]
#220 to be merged first :)

**Comment by mts299** [comment]
@carleyjmartin I just approved it for the sake of #220 being merged... but now you need to resolve some conflicts ;)

File: pydarn/plotting/fan.py [hunk 1]
```diff
@@ -177,15 +178,14 @@ def plot_fan(cls, dmap_data: List[dict], ax=None, ranges: List=[],
                                                          scan_time)
         # Locate scan in loaded data
         plot_beams = np.where(beam_scan == scan_index)
-        hemisphere = SuperDARNRadars.radars[dmap_data[0]['stid']].hemisphere
```
Comment (carleyjmartin) [inline pydarn/plotting/fan.py:179]: I can't plot southern hemisphere data any more, it plots the northern hemisphere with nothing on it. This change and the kwargs['hemisphere'] change below could be the culprit.

File: pydarn/plotting/fan.py [hunk 2]
```diff
@@ -271,9 +270,6 @@ def plot_fan(cls, dmap_data: List[dict], ax=None, ranges: List=[],
         rs = rs[0:ranges[1]-ranges[0]+1]
         scan = scan[0:ranges[1]-ranges[0]]
 
-        stid=dmap_data[0]['stid']
-        kwargs['hemisphere'] = SuperDARNRadars.radars[stid].hemisphere
```
Comment (carleyjmartin) [inline pydarn/plotting/fan.py:273]: Second place where southern hemisphere issue could stem from?

**Comment by mts299** [comment]
@carleyjmartin can you look into this and get it working?

**Comment by carleyjmartin** [comment]
Re-included the kwargs['hemisphere'] setter in the section that calls the axes for the first time and it appears to be working again.

Tested with:

``` python
pydarn.Fan.plot_fan(fitacf_data, grid=True, scan_index=2, fov_color='blue',
                    radar_label=True, groundscatter=True,ranges=[30,60])
```
and
```python                    
pydarn.Fan.plot_fan(fitacf_data, grid=True, scan_index=2, fov_color='blue',
                    radar_label=True, groundscatter=True,ranges=[30,60],coords=pydarn.Coords.GEOGRAPHIC ,projs=pydarn.Projs.GEO_COASTALINE)
 ```

For both DNC and RKN to test both hemispheres, and all appear as expected.
Let me know if there's anything else a bit off since the merge with cartopy!

**Comment by mts299** [comment]
@carleyjmartin if fix the conflicts I will re-test then hopefully merge?

**Comment by carleyjmartin** [comment]
I'm very surprised that there was only one basic conflict with develop, so definitely test again, and def test on a geographic plot if you can :P

**Comment by mts299** [comment]
@carleyjmartin south works but now breaks with the cartopy addition:

``` python3
import bz2 
import pydarn
import datetime as dt
import matplotlib.pyplot as plt 

data_file = "/home/marina/superdarn/data/fitacf/20220108.0801.00.dcn.fitacf.bz2" 
with bz2.open(data_file) as fp: 
    fitacf_stream = fp.read()

data = pydarn.SuperDARNRead(fitacf_stream, True).read_fitacf()

pydarn.Fan.plot_fan(data, grid=True, scan_index=5, ranges=[20,45], fov_color='blue', radar_label=True, groundscatter=True)
plt.show()
#
pydarn.Fan.plot_fan(data, grid=True, scan_index=5, fov_color='blue', radar_label=True, groundscatter=True, coords=pydarn.Coords.GEOGRAPHIC,  projs=pydarn.Projs.GEO_COASTALINE)
plt.show()

```

``` bash
python3 test_fan.py 
resetting environment variable IGRF_COEFFS in python script
resetting environment variable AACGM_v2_DAT_PREFIX in python script
non-default coefficient files may be specified by running aacgmv2.wrapper.set_coeff_path before any other functions

IMPORTANT: Please make sure to cite pyDARN in publications that use plots created by pyDARN using DOI: https://zenodo.org/record/3727269. Citing information for SuperDARN data is found at https://pydarn.readthedocs.io/en/master/user/citing/
Traceback (most recent call last):
  File "test_fan.py", line 22, in <module>
    pydarn.Fan.plot_fan(data, grid=True, scan_index=5, fov_color='blue', radar_label=True, groundscatter=True, coords=pydarn.Coords.GEOGRAPHIC, projs=pydarn.Projs.GEO_COASTALINE)
  File "/home/marina/.local/lib/python3.8/site-packages/pydarn/plotting/fan.py", line 303, in plot_fan
    ax.pcolormesh(np.degrees(thetas), rs,
  File "/home/marina/.local/lib/python3.8/site-packages/cartopy/mpl/geoaxes.py", line 321, in wrapper
    return func(self, *args, **kwargs)
  File "/home/marina/.local/lib/python3.8/site-packages/cartopy/mpl/geoaxes.py", line 1666, in pcolormesh
    result = self._pcolormesh_patched(*args, **kwargs)
  File "/home/marina/.local/lib/python3.8/site-packages/cartopy/mpl/geoaxes.py", line 1702, in _pcolormesh_patched
    X, Y, C, shading = self._pcolorargs('pcolormesh', *args,
  File "/home/marina/.local/lib/python3.8/site-packages/matplotlib/axes/_axes.py", line 5575, in _pcolorargs
    raise TypeError('Dimensions of C %s are incompatible with'
TypeError: Dimensions of C (76, 16) are incompatible with X (17) and/or Y (76); see help(pcolormesh)
```

**Comment by carleyjmartin** [comment]
I can't recreate the error you're finding, it's also occurring on a line that I haven't touched in this PR so it might be a bug somewhere else?

I've used:
```python
import matplotlib.pyplot as plt
import pydarn

file = '/Users/carley/Documents/data/20220108.0801.00.dcn.fitacf'
SDarn_read = pydarn.SuperDARNRead(file)
fitacf_data = SDarn_read.read_fitacf()

#pydarn.Fan.plot_fan(fitacf_data, grid=True, scan_index=2, fov_color='blue',
#                    radar_label=True, groundscatter=True, ranges=[30,60])
                    
pydarn.Fan.plot_fan(fitacf_data, grid=True, scan_index=5, fov_color='blue',
                    radar_label=True, groundscatter=True, ranges=[30,60],
                    coords=pydarn.Coords.GEOGRAPHIC ,projs=pydarn.Projs.GEO_COASTALINE)
    
plt.show()
```

and get 
Screen Shot 2022-03-03 at 9 32 39 AM

**Comment by mts299** [comment]
its weird because I only get this on this PR branch

**Pull request merged**